group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1053,6 +1055,12 @@ ConsumerGroup convertToConsumerGroup(ClassicGroup classicGroup, List<Coordinator
 
             throw new GroupIdNotFoundException("Cannot upgrade the classic group " + classicGroup.groupId() +
                 " to consumer group because the embedded consumer protocol is malformed.");
+        } catch (UnsupportedVersionException e) {
+            log.warn("Cannot upgrade the classic group " + classicGroup.groupId() +
+                " to consumer group: " + e.getMessage() + ".", e);
+
+            throw new GroupIdNotFoundException("Cannot upgrade the classic group " + classicGroup.groupId() +
+                " to consumer group because a custom assignor is in use.");
```

Comment:
Completely unrelated but I wonder if we always provide such good error message when the upgrade is not possible. Do you mind checking?

Reply:
nit: I wonder if we could extend the error message a bit more. We actually support upgrading while using a custom assignor but only if it does not have user data associated to the assignment. We could perhaps explain why they should do in this case (e.g. move to a default assignor and then upgrade?)?

Reply:
> Completely unrelated but I wonder if we always provide such good error message when the upgrade is not possible. Do you mind checking?

I had a look and there are three reasons in [`validateOnlineUpgrade`](https://github.com/apache/kafka/blob/e99561e1f30a30f9b151669a78d340f286a3ead6/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java#L1008-L1023) that we convert to "Group %s is not a consumer group":
 * Cannot upgrade classic group {} to consumer group because the online upgrade is disabled.
 * Cannot upgrade classic group {} to consumer group because the group does not use the consumer embedded protocol.
 * Cannot upgrade classic group {} to consumer group because the group size exceeds the consumer group maximum size.

Apart from that and SchemaException when subscriptions or metadata are malformed, I can't see any other error paths directly related to the upgrade.

Reply:
> I had a look and there are three reasons in [validateOnlineUpgrade](https://github.com/apache/kafka/blob/e99561e1f30a30f9b151669a78d340f286a3ead6/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java#L1008-L1023) that we convert to "Group %s is not a consumer group":

I wonder whether we should extend the error message that we provide to the user. "not a consumer group" does not say much. What do you think?

Reply:
I've updated `validateOnlineUpgrade` to throw an exception with the appropriate message instead. And tweaked some of the error wording to be consistent with the rest.